MetaMask/metamask-extension
# `Box` component `0` value fix

## Explanation

**What is the current state of things and why does it need to change?**
Currently `0` value option does not work on the `Box` component margin or padding props. The CSS styles exists but the classNames are not being generated. This was introduced during the responsive props update https://github.com/MetaMask/metamask-extension/pull/15106 and completely my fault 🤦 

**What is the solution your changes offer and how does it work?**
- The `Box` component logic has been updated to resolve the `0` value issue and now generates the correct class names
- Tests have been update to have a more comprehensive coverage for margin, padding and gap.
- Responsive story has been updated to show more responsive prop examples.
- Snapshots updated with missing margin/padding 0 class names

This update removes the `&&` operator during the `generateClassNames` call for each prop and instead does a value check inside of `generateClassNames`. This means `generateClassNames` is called regardless of it's prop value. I've run some basic performance tests and couldn't see a huge change but open to suggestions on this. `0.4ms` vs `0.466ms` after the update run in an incognito window using the react dev tools profiler have documented results below.

* Fixes #16976 

## Screenshots/Screencaps

### Before


https://user-images.githubusercontent.com/8112138/208551691-7e030506-39cc-4ec8-b06a-aa3fb0436311.mov

https://user-images.githubusercontent.com/8112138/208551721-2031caaf-e6ff-4029-a164-fe42b1134b7e.mov

#### Profiler testing performance
- Render 1: `0.4ms`
- Render 2: `0.3ms`
- Render 3: `0.5ms`

Average of `0.4ms`

<img width="721" alt="Screenshot 2022-12-19 at 4 25 03 PM" src="https://user-images.githubusercontent.com/8112138/208553598-4673fadd-f2d5-4df9-9e0e-24b78fac015b.png">
<img width="1439" alt="Screenshot 2022-12-19 at 4 25 40 PM" src="https://user-images.githubusercontent.com/8112138/208553600-ab3e9378-e90d-4c07-a86e-24fb2bb0d660.png">
<img width="1440" alt="Screenshot 2022-12-19 at 4 26 16 PM" src="https://user-images.githubusercontent.com/8112138/208553603-14452fd3-89a5-4328-8383-20c7b74b3afb.png">


### After


https://user-images.githubusercontent.com/8112138/208551746-5643e43a-f459-4f1e-bdf4-dcdd93380e40.mov

https://user-images.githubusercontent.com/8112138/208551766-f24ec980-4f3a-4ed5-ae61-5017cc714c72.mov

#### Profiler testing performance
- Render 1: `0.5ms`
- Render 2: `0.7ms`
- Render 3: `0.2ms`

Average of `0.466ms`

<img width="1440" alt="Screenshot 2022-12-19 at 4 22 39 PM" src="https://user-images.githubusercontent.com/8112138/208553617-46134ec3-242e-47cc-b30a-0a7b729aecc0.png">
<img width="1438" alt="Screenshot 2022-12-19 at 4 23 22 PM" src="https://user-images.githubusercontent.com/8112138/208553621-a7875af4-5935-4047-ade9-68b9d683114c.png">
<img width="1440" alt="Screenshot 2022-12-19 at 4 24 06 PM" src="https://user-images.githubusercontent.com/8112138/208553625-fc92c8a5-f93d-4845-ac4d-be9ef5f184c8.png">


## Manual Testing Steps

- Go to the latest storybook build of this PR
- Search `Box` in the search panel
- Update any of the bounding box controls (margin, marginTop, padding, paddingTop, etc) to use `0` value
- In developer tools check that the className and CSS have been added correctly

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
